spark/v3.4/spark/src/jmh/java/org/apache/iceberg/spark/source/WritersBenchmark.java

```diff
@@ -363,6 +389,60 @@ public void writeUnpartitionedClusteredPositionDeleteWriter(Blackhole blackhole)
     blackhole.consume(writer);
   }
 
+  @Benchmark
+  @Threads(1)
+  public void writeUnpartitionedFanoutPositionDeleteWriter(Blackhole blackhole) throws IOException {
+    FileIO io = table().io();
+
+    OutputFileFactory fileFactory = newFileFactory();
+    SparkFileWriterFactory writerFactory =
+        SparkFileWriterFactory.builderFor(table()).dataFileFormat(fileFormat()).build();
+
+    FanoutPositionOnlyDeleteWriter<InternalRow> writer =
+        new FanoutPositionOnlyDeleteWriter<>(
+            writerFactory, fileFactory, io, TARGET_FILE_SIZE_IN_BYTES);
+
+    PositionDelete<InternalRow> positionDelete = PositionDelete.create();
+    try (FanoutPositionOnlyDeleteWriter<InternalRow> closeableWriter = writer) {
+      for (InternalRow row : positionDeleteRows) {
+        String path = row.getString(0);
+        long pos = row.getLong(1);
+        positionDelete.set(path, pos, null);
+        closeableWriter.write(positionDelete, unpartitionedSpec, null);
+      }
+    }
+
+    blackhole.consume(writer);
+  }
+
+  @Benchmark
+  @Threads(1)
+  public void writeUnpartitionedFanoutPositionDeleteWriterShuffled(Blackhole blackhole)
```

Comment:
We should expect 5-15% overhead for the new buffering writer, which can still be beneficial for the job if we skip local ordering for inserts and potentially avoid spilling. This benchmark also does not take into account the cost to order records, it only tests the write performance. We will use this writer only if fanout is enabled. We should also explore Puffin delete files that would persist bitmaps directly.

```
Benchmark                                                                                                      Mode  Cnt           Score            Error   Units
ParquetWritersBenchmark.writeUnpartitionedClusteredPositionDeleteWriter                                          ss    5           6.004 ±          0.185    s/op
ParquetWritersBenchmark.writeUnpartitionedFanoutPositionDeleteWriter                                             ss    5           6.503 ±          0.171    s/op
ParquetWritersBenchmark.writeUnpartitionedFanoutPositionDeleteWriterShuffled                                     ss    5           6.616 ±          0.204    s/op
```

Reply:
> We should also explore Puffin delete files that would persist bitmaps directly

+1

Reply:
About memory overhead (not sure any thing measures it now):  Should be just additional space of map (data_file_path => bitmaps)?  Will there be cases, esp in fanout, where a writer writes many delete of many data files, that will start to stress it?

Reply:
I ran this benchmark (100 data files, 50k deletes each, 5 million deletes total) with a GC profiler and did not see anything bad. Issues will arise when there are lots of unique data files. That's unlikely as we distribute by partition and this writer will still be disabled by default, so users will have to opt in explicitly.